src/amo/sagas/collections.js

```diff
@@ -144,36 +169,40 @@ export function* updateCollection({
     formOverlayId,
     isPublic,
     name,
-    slug,
     user,
   },
-}) {
+}: UpdateCollectionAction): Generator<any, any, any> {
   const errorHandler = createErrorHandler(errorHandlerId);
   yield put(errorHandler.createClearingAction());
 
   try {
+    yield put(beginFormOverlaySubmit(formOverlayId));
+
     const state = yield select(getState);
-    yield call(api.updateCollection, {
+    const params: UpdateCollectionParams = {
       api: state.api,
       collectionSlug,
       defaultLocale,
       description,
       isPublic,
       name,
-      slug,
       user,
-    });
+    };
+    yield call(api.updateCollection, params);
 
+    // Invalidate the stored collection object. This will force each
+    // component to re-fetch the collection.
+    yield put(deleteCollectionBySlug(collectionSlug));
     yield put(closeFormOverlay(formOverlayId));
-    yield put(finishUpdateCollection({ collectionSlug, successful: true }));
+    yield put(finishFormOverlaySubmit(formOverlayId));
```

Comment:
Seeing this in the collections saga makes me feel a bit uneasy though. I get it's more generic/reusable but it feels like collections saga/API work shouldn't be so tightly coupled to the UI implementation of a form overlay. What happens if this is done elsewhere without a form overlay?

I know it was like that before and it's the only place it's used now... and also I guess it wouldn't be hard to abstract out so it's fine. I guess it just seem jarring to me to have the update action in the collection saga be fired by a UI state/action. It feels a bit too tightly coupled to me.

Reply:
Yeah, I definitely hear what you're saying and this was on my mind as well.

> What happens if this is done elsewhere without a form overlay?

My current idea is that we would make another saga that shares some of the same update code -- i.e. we'd have two sagas like `updateCollectionFromOverlay` and `updateCollectionFromPage`. This will be easy to do because a saga can dispatch an action to trigger another saga.

But, yeah, I'm not sure if this is the best pattern. To be fair, a saga really is meant to encapsulate a series of actions so maybe it makes sense.

I think we should keep iterating and see what happens. One thing worth mentioning is that we have no intention of hosting an edit collection screen *not* in an overlay 😄 so this may never be an issue.

If you have any other ideas for how to implement it, let me know. My original approach was to make the begin/finish state more specific to the act of *editing a collection* but then I had to infer the form overlay state by inspecting the collection state so it just felt like saving state in two places whereas this patch does it in one.